vassal-app/src/main/java/VASSAL/build/module/map/PieceMover.java

```diff
@@ -806,6 +809,58 @@ public Command movePieces(Map map, Point p) {
       // Any piece we successfully moved, make sure is now considered a "selected piece" (i.e. a member of KeyBuffer)
       for (final GamePiece piece : draggedPieces) {
         KeyBuffer.getBuffer().add(piece);
+
+        // If this is a piece that can be placed on mats, look for an overlapping mat, and put it there.
+        if (GameModule.getGameModule().isMatSupport()) {
+          if ("true".equals(piece.getProperty(MatCargo.IS_CARGO))) { //NON-NLS
+            final MatCargo cargo = (MatCargo)Decorator.getDecorator(piece, MatCargo.class);
+            final GamePiece oldMat = cargo.getMat();
+
+            if ((oldMat == null) || (!draggedPieces.contains(oldMat) && !allDraggedPieces.contains(oldMat) && !DragBuffer.getBuffer().contains(oldMat))) {
+              final GamePiece newMat = map.findAnyPiece(p, PieceFinder.MAT_ONLY);
+              if (newMat != null) {
+                final Mat mat = (Mat) Decorator.getDecorator(newMat, Mat.class);
+                if (mat != null) {
+                  comm = comm.append(mat.makeAddCargoCommand(piece));
+                }
+              }
+              else {
+                // We're NOT on a mat, so if we WERE on one, mark ourselves as no longer on it
+                comm = comm.append(cargo.makeClearMatCommand());
+              }
+            }
+          }
+          else {
```

Comment:
I tried various options of unselecting cargo and moving the mat out from under it, then back, etc. all worked as I expected. The basic functionality looks great.